lib/trino-cache/src/main/java/io/trino/cache/EvictableCache.java

```diff
@@ -78,13 +78,15 @@ class EvictableCache<K, V>
                         .<Token<K>, V>removalListener(removal -> {
                             Token<K> token = removal.getKey();
                             verify(token != null, "token is null");
-                            if (removal.getCause() == RemovalCause.REPLACED) {
-                                return;
+                            synchronized (token) {
```

Comment:
Document why this is needed, what are the invariants that require synchronization.
Same in other places where synchronization has been added.